### PR TITLE
Rewrite var clushing spark

### PR DESCRIPTION
**Pull request by ZhuLi99**

### CODE REVIEW

File: src/main/anovos/data_analyzer/association_eval_varclus.py [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+        for _ in range(n_rs):
+            random.shuffle(feat_list)
+            rs_clus1, rs_clus2, rs_var = self._reassign(df, clus1, clus2, feat_list)
+            if rs_var > max_rs_var:
+                max_rs_var = rs_var
+                fin_rs_clus1, fin_rs_clus2 = rs_clus1, rs_clus2
+
+        return fin_rs_clus1, fin_rs_clus2, max_rs_var
+
+    def _varclusspu(self):
```
Comment (dattranm) [inline src/main/anovos/data_analyzer/association_eval_varclus.py:216]: As above, we can include `spark` argument in here, instead of in `init`. 

One more thing, _varclusspu is named by the VarClusHi author, stands for "speed-up" version. We can rename this here, to something that makes more sense and distinguish our function, something like `_varclusspark`.

File: src/main/anovos/data_analyzer/association_eval_varclus.py [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+        """
+        This function is the main function which performs variable clustering.
+        By default, it begins with all variables in a single cluster. It then repeats the following steps:
+            1. A cluster is chosen for splitting. This cluster has the largest eigenvalue associated with the 2nd PC
+            2. Find the first 2 PCs of the chosen cluster  split into 2 clusters perform an orthoblique rotation (raw quartimax rotation on the eigenvectors)
```
Comment (dattranm) [inline src/main/anovos/data_analyzer/association_eval_varclus.py:221]: Check the font of the docstring here, it has some font errors

File: src/main/anovos/data_analyzer/association_eval_varclus.py [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+            3. Variables are iteratively re-assigned to clusters to maximize the variance
+                a. Nearest Component Sorting (NCS) Phase: In each iteration, the cluster components are computed. Each variable is assigned to the rotated component with which it has the highest squared correlation
+                b. Search Phase: Assign each variable into different cluster  see if this increase the variance explained. If variable is re-assigned here, the components of the clusters involved are recomputed before next variable is tested
```
Comment (dattranm) [inline src/main/anovos/data_analyzer/association_eval_varclus.py:224]: Same as above, font shows  .

File: src/main/anovos/data_analyzer/association_evaluator.py [hunk 4]
```diff
@@ -146,17 +147,15 @@ def variable_clustering(
     idf,
     list_of_cols="all",
     drop_cols=[],
-    sample_size=100000,
     stats_mode={},
     print_impact=False,
 ):
     """
     Variable Clustering groups attributes that are as correlated as possible among themselves within a cluster and
```
Comment (dattranm) [inline src/main/anovos/data_analyzer/association_evaluator.py:154]: We can say in here that variable_clustering is a wrapper for our VarClusHiSpark, and briefly introduce the function and algorithm more.

File: src/main/anovos/data_analyzer/association_evaluator.py [hunk 5]
```diff
@@ -146,17 +147,15 @@ def variable_clustering(
...
-    to do variable clustering; however, this library is not implemented in a scalable manner due to which the
-    analysis is done on a sample dataset. Further, it is found to be a bit computational expensive especially when
-    number of columns in the input dataset is on higher side (number of pairs to analyse increases exponentially with
-    number of columns).
+    as uncorrelated as possible with attribute in other clusters. The function is leveraging VarClusHiSpark class from
+    anovos.data_analyzer.association_eval_varclus file, which is a scalable version of [VarClusHi] [2] library
+    to do variable clustering. The VarClusHiSpark is implemented in a scalble manner with necessary optimizations,
+    thus sampling is not required.
 
     [2]: https://github.com/jingtt/varclushi   "VarCluShi"
```
Comment (dattranm) [inline src/main/anovos/data_analyzer/association_evaluator.py:160]: We can move this, and the explanation for "scalable version of VarClusHi" inside our `VarClusHiSpark` as stated above, and only focus on briefly walking through the algorithm and the wrapper logic here.

File: src/main/anovos/data_analyzer/association_evaluator.py [hunk 6]
```diff
@@ -215,40 +214,32 @@ def variable_clustering(
...
 
-    idf_sample = data_sample(
-        idf,
-        fraction=min(1.0, float(sample_size) / idf.count()),
-        method_type="random",
-        seed_value=0,
-    )
-    idf_sample.persist(pyspark.StorageLevel.MEMORY_AND_DISK).count()
-
+    idf.persist(pyspark.StorageLevel.MEMORY_AND_DISK).count()
```
Comment (dattranm) [inline src/main/anovos/data_analyzer/association_evaluator.py:217]: For `persist`, we should create an argument and let the user choose on whether to persist or not, and their preferred persist option. For reference, check out https://github.com/anovos/anovos/blob/e0b3055a33415a211911744efbacafd0f100ad38/src/main/anovos/drift_stability/drift_detector.py#L147

File: src/main/anovos/data_analyzer/association_evaluator.py [hunk 7]
```diff
@@ -215,40 +214,32 @@ def variable_clustering(
...
-            idf_sample = idf_sample.withColumn(i[0], F.col(i[0]).cast("double"))
+            idf = idf.withColumn(i[0], F.col(i[0]).cast("double"))
     idf_encoded = cat_to_num_unsupervised(
-        spark, idf_sample, list_of_cols=cat_cols, method_type="label_encoding"
+        spark, idf, list_of_cols=cat_cols, method_type="label_encoding"
     )
     num_cols = attributeType_segregation(idf_encoded)[0]
     idf_encoded = idf_encoded.select(num_cols)
     idf_imputed = imputation_MMM(spark, idf_encoded, stats_mode=stats_mode)
     idf_imputed.persist(pyspark.StorageLevel.MEMORY_AND_DISK).count()
```
Comment (dattranm) [inline src/main/anovos/data_analyzer/association_evaluator.py:238]: Same as above, put `persist` under an argument

File: src/main/anovos/data_analyzer/association_evaluator.py [hunk 8]
```diff
@@ -215,40 +214,32 @@ def variable_clustering(
...
-    idf_sample.unpersist()
-    idf_pd = idf_imputed.toPandas()
-    vc = VarClusHi(idf_pd, maxeigval2=1, maxclus=None)
-    vc.varclus()
-    odf_pd = vc.rsquare
+    idf.unpersist()
```
Comment (dattranm) [inline src/main/anovos/data_analyzer/association_evaluator.py:239]: Same as above, put persist under an argument

File: src/main/anovos/data_analyzer/association_evaluator.py [hunk 9]
```diff
@@ -215,40 +214,32 @@ def variable_clustering(
...
+    vc = VarClusHiSpark(spark, idf_imputed, maxeigval2=1, maxclus=None)
+    vc._varclusspu()
+    odf_pd = vc._rsquarespu()
```
Comment (dattranm) [inline src/main/anovos/data_analyzer/association_evaluator.py:242]: Need to `unpersist` `idf_imputed` before the function ends to avoid overflow

**Comment by ZhuLi99** [comment]
> Thank @ZhuLi99 for the PR! I have some points that need to be addressed before approving and merging.
> 
> Most logic for `VarClusHiSpark` has been tested and confirmed working properly and retained the same result as intended.
> 
> One more thing, if possible, can you add an additional test case for `variable_clustering` for the null bug that we encountered recently?

Hi @dattranm I have added this [here](https://github.com/anovos/anovos/blob/main/src/test/anovos/data_analyzer/test_association_evaluator.py#L272) and this branch uses the same [test](https://github.com/anovos/anovos/blob/rewrite-var-clushing-spark/src/test/anovos/data_analyzer/test_association_evaluator.py#L272).

File: src/main/anovos/data_analyzer/association_eval_varclus.py [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+    """
+
+    def __init__(self, df, feat_list=None, maxeigval2=1, maxclus=None, n_rs=0):
+        """
+        Parameters
+        ----------
+        df
+            PySpark Dataframe
+        feat_list
+            List of features to perform variable clustering e.g., ["col1","col2"].
```
Comment (dattranm) [inline src/main/anovos/data_analyzer/association_eval_varclus.py:40]: For all arguments that have a default value, can you put (Default value = ....) instead of e.g.? For example, for this one: (Default value = None). You can refer to here https://github.com/anovos/anovos/blob/e0b3055a33415a211911744efbacafd0f100ad38/src/main/anovos/data_analyzer/association_evaluator.py#L79

**Review by dattranm — CHANGES_REQUESTED** [review]

**Review by dattranm — APPROVED** [review]
All the reviews has been addressed. Approve this PR on my side. Thanks @ZhuLi99 .

**Review by VishnuGowthem — APPROVED** [review]
For later : we should look to add more test cases for src/main/anovos/data_analyzer/association_eval_varclus.py if possible in future updations.
